src/block/block_ext.c

```diff
@@ -999,23 +999,41 @@ __block_append(
      * object in the skiplist, check for a simple extension, and otherwise append a new structure.
      */
     if ((ext = el->last) != NULL && ext->off + ext->size == off)
+        /* ext is adjacent to off/size. */
         ext->size += size;
     else {
         ext = __block_off_srch_last(el->off, astack);
         if (ext != NULL && ext->off + ext->size == off)
+            /* ext is adjacent to off/size. */
             ext->size += size;
-        else {
+        else if (ext == NULL) {
             WT_RET(__wti_block_ext_alloc(session, &ext));
             ext->off = off;
             ext->size = size;
 
             for (i = 0; i < ext->depth; ++i)
                 *astack[i] = ext;
             ++el->entries;
-        }
 
-        /* Update the cached end-of-list */
-        el->last = ext;
+            /* Update the cached end-of-list */
+            el->last = ext;
+        } else if (ext->off >= (off + size)) {
```

Comment:
So this is append to the front semantics? Which looks correct to me but then also duplicates a bit of code above. Can't think of a great way to avoid that duplication.

Reply:
Append to the end sematics.
See diagrams below that show points in order.
* Existing extent list is empty -- Worked before this change.
* Adjacent to the end of the last extent: off = ext->off + ext->size: * 0, * ext->off,  * ext->off + ext->size and off -- Worked before this change
* Past the end of the last extent: ext->off + ext->size < off: * 0, * ext->off, * ext->off + ext->size, gap, * off -- Did not work since it did not set next.
* Before the last extent: off + size < ext->off: * 0, * off, * off + size, gap, * ext->off -- Did not work. Not an append. Requires __block_merge()
* Overlaps the last extent: off < ext->off + ext->size && off + size > ext->off: * off or ext->off, * ext->off + ext->size, gap, ext->off + ext->size

Reply:
> Before the last extent: off + size < ext->off: * 0, * off, * off + size, gap, * ext->off -- Did not work. Not an append. Requires __block_merge().
>Overlaps the last extent: off < ext->off + ext->size && ext->off + ext->size < off + size: * off or ext->off, * ext->off or off, * ext->off + ext->size, gap, off + size -- Did not work. Not an append. Requires __block_merge().

Should these cases then be assertions? If someone was to call append but not provide appendable offsets then maybe that is a bug.

Reply:
Using an assert might be OK.  The comment in __block_append implies it should only be called when an append is happening.

```
   /*
     * Identical to __block_merge, when we know the file is being extended, that is, the information
     * is either going to be used to extend the last object on the list, or become a new object
     * ending the list.
...
     */
```

That comment above gave me an idea.  I should comment the code branches with the same wording and terminology in that comment.

Never-the-less, I would rather write forgiving code especially since we already have the code to handle that.

I could run a full patch build replacing the call to __block_merge() with an assert.  That would prove whether doing that is OK.

Reply:
This is an append function if we are not appending I think we should at least try to assert.

Reply:
I have added a WT_ASSERT().  However, I think it is better to leave it out.  I call __block_merge() to handle the case of adding a new extent after but not adjacent to the last extent.  As long as we are doing that, why not just do it for all cases not covered by __block_append() itself?

Reply:
This is a function that is meant to append, it is a bit weird if it handles non append cases without any warning. Perhaps instead of an assert it should return EINVAL? or WT_ERROR?

Reply:
Keep it as an assert for now, that way it won't generate a new error code path. I am still of the opinion that if we expect something to be an append that we ensure it's an append.